### DISCUSSION TITLE
Why isn't beartype complaining with callables?

### DISCUSSION THREAD

**Discussion by Artur-Galstyan** [beartype/beartype#409] [Q&A]
Why isn't beartype complaining here? (Pyright is, but I want beartype to complain too!)

```python
from beartype import beartype
from beartype.typing import Callable
from jaxtyping import Array, PyTree

class RootFnOutput:
    params: PyTree
    value: Array
    state: PyTree


RootFnCallable = Callable[..., RootFnOutput]


def get_root() -> int:
    return 123


@beartype
def forward(root_fn: RootFnCallable):
    root_fn()


if __name__ == "__main__":
    forward(get_root) # why u no complain?? :(
```

**Comment by leycec** [comment]
*...heh.* The answer to all these questions (and more) is: "...because @leycec is lazy." Sometimes, it's not a virtue. :cry: 

In other words, @beartype currently only **shallowly type-checks callables.** Clearly, this is non-ideal. Clearly, however, deeply type-checking callables at runtime in the way that you and everybody else wants is *super* non-trivial. In theory, it's feasible for various subsets of pure-Python callables; in practice, it's rough stuff that we'd better think about very carefully before even attempting.

See also [the official @beartype feature support matrix](https://beartype.readthedocs.io/en/latest/pep/). `Callable`, for example, is still listed as "*none*" under **full support** – meaning that @beartype currently only shallowly type-checks callables. Pour one out for @beartype. :sob:

**Reply by Artur-Galstyan** [reply]
Ah okay thanks for the answer! I ended up using a protocol class instead - this gets the job done too and beartype works as intended!

**Comment by leycec** [comment]
**Yay!** So sorry about that, @Artur-Galstyan. And thanks for sticking it out with @beartype despite API growing pains like this. :disappointed: 

Proper `Callable[...]` type-checking is *definitely* on the roster for an upcoming release – but leaning towards the end of the year now. The future is so bright I can barely see it.

**Reply by Artur-Galstyan** [reply]
No worries! It's beartype until the bitter end for me!
